### PR TITLE
LPS-145156 & LPS-145158 & LPS-140173  [CP 2.0] Developer Keys for DXP and DXP Cloud

### PR DESCRIPTION
**Pull request by Lucas2691**
We made Developer Keys component for the customer download a key from DXP and DXP Cloud.
To do this, we implemented a generic component to use on two story's.
The Developer Keys download a XML file to DXP version what customer choose.

To build a component:
We used "React" to request from database the DXP versions that is possible to choose on select field and download the keys.
If the project have DXP version, that version will be the default option.

To access and download the XML file on DXP:
- On Customer Portal, select a project what you want download a key for DXP (that project need to have DXP active).
- Click on product DXP on Side Menu (Product Activation).
- Select the version from your DXP.
- Click on button "Download Key" to download the Key (XML file).

To access and download the XML file on DXP Cloud:
- On Customer Portal, select a project what you want download a key for DXP Cloud (that project need to have DXP Cloud active).
- Click on product DXP Cloud on Side Menu (Product Activation).
- Select the version from your DXP Cloud.
- Click on button "Download Key" to download the Key (XML file).

### CODE REVIEW

**Comment by Lucas2691** [comment]
@igordmf

File: modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	productTitle,
+	sessionId,
+}) => {
+	const {licenseKeyDownloadURL} = useApplicationProvider();
+	const [dxpVersions, setDxpVersions] = useState([]);
+
+	const [selectedVersion, setSelectedVersion] = useState(dxpVersion);
+
+	const [hasLicenseDownloadError, setLicenseDownloadError] = useState(false);
+
```
Comment (jairmedeiros) [inline modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js:42]:
```suggestion
	const [selectedVersion, setSelectedVersion] = useState(dxpVersion);
	const [hasLicenseDownloadError, setLicenseDownloadError] = useState(false);
```

File: modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	const downloadText = DeveloperKeysText[productTitle];
+
+	useEffect(() => {
+		const fetchListTypeDefinitions = async () => {
+			const {data} = await client.query({
+				query: getListTypeDefinitions,
+				variables: {
+					filter: `name eq 'DXP Version'`,
```
Comment (jairmedeiros) [inline modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js:50]: Try to pass this 'DXP Version' to a const variable outside the component as `DEVELOPER_KEYS_HELPER`
```suggestion
					filter: `name eq '${LIST_TYPES.dxpVersion}'`,
```

File: modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+			setLicenseDownloadError(true);
+
+			return downloadFromBlob(licenseBlob, `license${extensionFile}`);
+		}
+	};
+
+	return (
+		<div>
+			<p className="text-neutral-7 text-paragraph">{downloadText}</p>
```
Comment (jairmedeiros) [inline modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js:92]:
```suggestion
			<p className="text-neutral-7 text-paragraph">{downloadTextHelper}</p>
```

File: modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+import {useEffect, useState} from 'react';
+import client from '../../../../../apolloClient';
+import BaseButton from '../../../../../common/components/BaseButton';
+import {useApplicationProvider} from '../../../../../common/context/ApplicationPropertiesProvider';
+import {getListTypeDefinitions} from '../../../../../common/services/liferay/graphql/queries';
+import {fetchDeveloperKeysLicense} from '../../../../../common/services/liferay/raysource-api';
+import {downloadFromBlob} from '../../../../../common/utils';
+import {EXTENSIONS_FILE_TYPE, STATUS_CODE} from '../../../utils/constants';
+
+const DeveloperKeysText = {
```
Comment (jairmedeiros) [inline modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js:23]:
```suggestion
const DEVELOPER_KEYS_HELPER = {
```

File: modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js (hunk 1 above)
Comment (Lucas2691) [reply modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js:42]: The Prettier do that when I save.

File: modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+			const {data} = await client.query({
+				query: getListTypeDefinitions,
+				variables: {
+					filter: `name eq 'DXP Version'`,
+				},
+			});
+
+			if (data) {
+				const items = data.listTypeDefinitions.items[0].listTypeEntries;
+				const orderedItems = [...items].sort((a, b) => b.name - a.name);
```
[the remark this answers is not in the document]
Comment (Lucas2691) [reply modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js:56]: Ok

File: modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+	sessionId,
+}) => {
+	const {licenseKeyDownloadURL} = useApplicationProvider();
+	const [dxpVersions, setDxpVersions] = useState([]);
+
+	const [selectedVersion, setSelectedVersion] = useState(dxpVersion);
+
+	const [hasLicenseDownloadError, setLicenseDownloadError] = useState(false);
+
+	const downloadText = DeveloperKeysText[productTitle];
```
[the remark this answers is not in the document]
Comment (Lucas2691) [reply modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js:43]: Ok

File: modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js (hunk 3 above)
Comment (Lucas2691) [reply modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js:92]: Ok

File: modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+					type="button"
+				>
+					Download Key
+				</BaseButton>
+
+				{hasLicenseDownloadError && (
+					<p className="mt-3 text-neutral-7 text-paragraph">
+						{`The requested activation key is not yet available. For more
+					information about the availability of your Enterprise Search
+					activation keys, please `}{' '}
```
[the remark this answers is not in the document]
Comment (Lucas2691) [reply modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js:140]: Ok

File: modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+			<p className="text-neutral-7">
+				For instructions on how to activate your Liferay DXP or Liferay
+				Portal instance, please read the
+				<a
+					href="https://help.liferay.com/hc/en-us/articles/360018163571-Deploying-Activation-Keys "
+					rel="noreferrer noopener"
+					target="_blank"
+				>
+					<u className="font-weight-semi-bold text-neutral-7">
+						{' '}
```
[the remark this answers is not in the document]
Comment (Lucas2691) [reply modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js:154]: Ok

File: modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+
+			if (data) {
+				const items = data.listTypeDefinitions.items[0].listTypeEntries;
+				const orderedItems = [...items].sort((a, b) => b.name - a.name);
+
+				setDxpVersions(orderedItems);
+
+				setSelectedVersion(
+					orderedItems.find((item) => item.name === dxpVersion)
+						?.name || orderedItems[0].name
```
[the remark this answers is not in the document]
Comment (Lucas2691) [reply modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js:62]: Ok

File: modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js (hunk 4 above)
Comment (Lucas2691) [reply modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js:23]: Ok

File: modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js (hunk 2 above)
Comment (Lucas2691) [reply modules/dxp/apps/osb/osb-site-initializer/osb-site-initializer-customer-portal/extra/remote-app/src/routes/customer-portal/components/DeveloperKeysLayout/Inputs/index.js:50]: Ok

**Comment by igordmf** [comment]
> @igordmf

ok

**Comment by jairmedeiros** [comment]
Just starting reviewing :)

**Comment by jairmedeiros** [comment]
resent https://github.com/pt-liferay-solutions/liferay-portal/pull/735

**Pull request closed without merging**